sw/device/tests/sim_dv/spi_tpm_smoketest.c

```diff
@@ -0,0 +1,217 @@
+      kDifSpiDeviceIrqUploadPayloadOverflow,
+      kDifSpiDeviceIrqReadbufWatermark,
+      kDifSpiDeviceIrqReadbufFlip,
+      kDifSpiDeviceIrqTpmHeaderNotEmpty};
+
+  for (uint32_t i = 0; i <= ARRAYSIZE(spi_device_irqs); ++i) {
+    CHECK_DIF_OK(dif_spi_device_irq_set_enabled(spi_device, spi_device_irqs[i],
+                                                kDifToggleEnabled));
+  }
+}
+
+void ottf_external_isr(void) {
+  plic_isr_ctx_t plic_ctx = {.rv_plic = &plic,
+                             .hart_id = kTopEarlgreyPlicTargetIbex0};
+
+  spi_device_isr_ctx_t spi_device_ctx = {
+      .spi_device = &spi_device.dev,
+      .plic_spi_device_start_irq_id =
+          kTopEarlgreyPlicIrqIdSpiDeviceGenericRxFull,
+      .expected_irq = kDifSpiDeviceIrqTpmHeaderNotEmpty,
+      .is_only_irq = true};
+
+  top_earlgrey_plic_peripheral_t peripheral;
+  dif_spi_device_irq_t spi_device_irq;
+  isr_testutils_spi_device_isr(plic_ctx, spi_device_ctx, &peripheral,
+                               &spi_device_irq);
+
+  switch (spi_device_irq) {
+    case kDifSpiDeviceIrqTpmHeaderNotEmpty:
+      header_interrupt_received = true;
+      // Disable interrupt until work is handled.
+      CHECK_DIF_OK(dif_spi_device_irq_set_enabled(
+          &spi_device.dev, kDifSpiDeviceIrqTpmHeaderNotEmpty,
+          kDifToggleDisabled));
+      break;
+    default:
+      LOG_ERROR("Unexpected interrupt: %d", spi_device_irq);
+      break;
+  }
+}
+
+static void ack_spi_tpm_header_irq(dif_spi_device_handle_t *spi_device) {
+  // Clear interrupt state and re-enable interrupt.
+  CHECK_DIF_OK(dif_spi_device_irq_acknowledge(
+      &spi_device->dev, kDifSpiDeviceIrqTpmHeaderNotEmpty));
+  CHECK_DIF_OK(dif_spi_device_irq_set_enabled(
+      &spi_device->dev, kDifSpiDeviceIrqTpmHeaderNotEmpty, kDifToggleEnabled));
+}
+
+bool test_main(void) {
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+
+  CHECK_DIF_OK(dif_spi_device_init_handle(
+      mmio_region_from_addr(TOP_EARLGREY_SPI_DEVICE_BASE_ADDR), &spi_device));
+
+  CHECK_DIF_OK(dif_rv_plic_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR), &plic));
+
+  // Set IoA7 for tpm csb.
+  // Longer term this needs to migrate to a top specific, platform specific
+  // setting.
+  CHECK_DIF_OK(dif_pinmux_input_select(
+      &pinmux, kTopEarlgreyPinmuxPeripheralInSpiDeviceTpmCsb,
+      kTopEarlgreyPinmuxInselIoa7));
+
+  CHECK_DIF_OK(
+      dif_spi_device_tpm_configure(&spi_device, kDifToggleEnabled, tpm_config));
+
+  // enable interrupts
+  en_plic_irqs(&plic);
+  en_spi_device_irqs(&spi_device.dev);
+
+  // Sync message with testbench to begin.
+  LOG_INFO("Begin TPM Test");
+
+  for (uint32_t i = 0; i < kIterations; i++) {
+    LOG_INFO("Iteration %d", i);
+
+    // Wait for write interrupt.
+    while (!header_interrupt_received) {
+      wait_for_interrupt();
+    }
+    header_interrupt_received = false;
```

Comment:
ah yes that's what i originally had.
the problem in this test case is that the SV side can run faster than the C side.
So if i don't have the check for example, the interrupt can actually happen "before" wait_for_interrupt is called, and the test just ends up getting stuck. 

The check is really meant to cover that case. It's not...100% proof, because an interrupt can still come in between the check and the wait for interrupt.

Reply:
Let me restructure this test a bit differently. I'll break the linear sequence and have everything just react on the write / read command.  Once we've handled N iterations of that, then the test can end.  That might it better without these loops and closer to how a real system would use it.

Reply:
I see. SV indeed could run faster than the C side in this case, unless we add more synchronizer for them.
Alex's suggestion may work.

Or we could add some comment to explain the rationale. Either way is fine to me.

Reply:
i've updated this portion and left a comment!